Issue806 Single zone VAV AHU economizer control sequence

milicag opened the pull request:
This PR adds a single zone VAV AHU economizer control sequence as prescribed by ASHRAE Guideline 36 (G36). The PR comprises 3 atomic and 1 composite control sequence models with the corresponding tests, scripts and validation results.

Atomic control sequences:

- ```EconDamperPositionLimitsSingleZone```: Outdoor and return air damper position limits for minimum outdoor airflow control
- ```EconEnableDisableSingleZone```: Economizer enable/disable switch
- ```EconModulationSingleZone```: Outdoor and return air damper position modulation for economizer operation

Composite control sequence:

- ```EconomizerSingleZone```: Single zone VAV AHU economizer control sequence.

It also includes improvements to the multizone economizer sequnece, such as better assignment of keywords ```final``` and ```protected```, several variable or block names, but also harmonizes single and multiple zone sequences. This makes validation more robust as well. In addition, the PR adds a default component name for the ```TrueFalseHold```, which was previously ommited.

milicag commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconEnableDisableMultiZone.mo:
```diff
@@ -4,83 +4,84 @@ block EconEnableDisableMultiZone
+  final parameter Modelica.SIunits.SpecificEnergy hOutHigLimCutHig = 0
     "Hysteresis block high limit cutoff";
-  parameter Real uEntHigLimCutLow(final unit="J/kg", quantity="SpecificEnergy") = uEntHigLimCutHig - delEntHis
+  final parameter Real hOutHigLimCutLow = hOutHigLimCutHig - delEntHis
```
These low limits need to be declared as ```Real``` as opposed to using ```Modelica.SIunits.*``` due to their negative values, which allow the sequence to be used both for differential and fixed cutoff temperature/enthalpy. Otherwise the inbuilt limits need to be breached and warnings get issued.

milicag commented:
@mwetter This is ready for first review.

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconDamperPositionLimitsMultiZone.mo:
```diff
@@ -36,23 +36,23 @@ block EconDamperPositionLimitsMultiZone
 
-  CDL.Interfaces.RealOutput yOutDamPosMin(min=0, max=1, unit="1")
+  CDL.Interfaces.RealOutput yOutDamPosMin(final min=outDamPhyPosMin, final max=outDamPhyPosMax, final unit="1")
     "Minimum outdoor air damper position limit"
```
@milicag it would be more readable to make each assignment occupied each line:
`final min=outDamPhyPosMin, `
`final max=outDamPhyPosMax, `
 `final unit="1"`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconDamperPositionLimitsSingleZone.mo:
```diff
@@ -0,0 +1,345 @@
+block EconDamperPositionLimitsSingleZone
+  "Single zone VAV AHU minimum outdoor air control - damper position limits"
+
+  parameter Real minFanSpe(final min=0, final max=1, final unit="1") = 0.1 "Minimum supply fan operation speed";
```
@milicag move the comments to a new line

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconEnableDisableMultiZone.mo:
```diff
@@ -4,83 +4,84 @@ block EconEnableDisableMultiZone
         extent={{180,-220},{200,-200}}), iconTransformation(extent={{100,-40},{140,0}})));
 
-  CDL.Logical.TrueFalseHold TrueFalseHold(duration=600) "10 min on/off delay"
+  CDL.Logical.TrueFalseHold trueFalseHold(final duration=600) "10 min on/off delay"
```
@milicag `trueFalseHold` to be `truFalHol`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/EconEnableDisableMultiZone.mo:
```diff
@@ -93,24 +94,24 @@ protected
     "Outdoor air temperature hysteresis for both fixed and differential dry bulb temperature cutoff conditions"
     annotation (Placement(transformation(extent={{-100,240},{-80,260}})));
-  CDL.Logical.Hysteresis hysOutEnt(final uLow=uEntHigLimCutLow, final uHigh=uEntHigLimCutHig) if use_enthalpy
+  CDL.Logical.Hysteresis hysOutEnt(final uLow=hOutHigLimCutLow, final uHigh=hOutHigLimCutHig) if use_enthalpy
```
@milicag each assignment occupies one line.

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/Validation/EconDamperPositionLimitsSingleZone_FanSpe_VOut_flow.mo:
```diff
@@ -0,0 +1,129 @@
+equation
+  connect(freProSta.y, ecoDamLim.uFreProSta)
+    annotation (Line(points={{-99,-90},{-60,-90},{-60,-18},{-41,-18}},
+                                                                   color={255,127,0}));
```
@milicag  remove the blank spaces

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/Validation/EconEnableDisableMultiZone_TOut_hOut.mo:
```diff
@@ -43,26 +28,41 @@ model EconEnableDisableMultiZone_TOut_hOut
+  CDL.Continuous.Sources.Constant retDamPhyPosMax(final k=1) "Maximal allowed economizer damper position"
     annotation (Placement(transformation(extent={{-160,-120},{-140,-100}})));
-  CDL.Integers.Sources.Constant ZoneState(k=Constants.ZoneStates.deadband) "Zone State is deadband"
+  CDL.Integers.Sources.Constant ZoneState(final k=Constants.ZoneStates.deadband) "Zone State is deadband"
```
@milicag `ZoneState` to be `zonSta`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/Validation/EconEnableDisableMultiZone_TOut_hOut.mo:
```diff
@@ -43,26 +28,41 @@ model EconEnableDisableMultiZone_TOut_hOut
+  CDL.Integers.Sources.Constant ZoneState(final k=Constants.ZoneStates.deadband) "Zone State is deadband"
     annotation (Placement(transformation(extent={{-200,-10},{-180,10}})));
-  CDL.Integers.Sources.Constant FreProSta(k=Constants.FreezeProtectionStages.stage0) "Freeze Protection Status - Disabled"
+  CDL.Integers.Sources.Constant FreProSta(final k=Constants.FreezeProtectionStages.stage0) "Freeze Protection Status - Disabled"
```
`FreProSta` to be `freProSta`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/Validation/EconEnableDisableMultiZone_TOut_hOut.mo:
```diff
@@ -43,26 +28,41 @@ model EconEnableDisableMultiZone_TOut_hOut
+  CDL.Integers.Sources.Constant FreProSta(final k=Constants.FreezeProtectionStages.stage0) "Freeze Protection Status - Disabled"
     annotation (Placement(transformation(extent={{-200,20},{-180,40}})));
-  CDL.Logical.Sources.Pulse booPul(final startTime=10, period=2000) "Boolean pulse signal"
+  CDL.Logical.Sources.Constant SupFanSta(final k=true) "Supply fan status signal"
```
`SupFanSta` to be `supFanSta`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/Validation/EconEnableDisableSingleZone_FreProSta_ZonSta.mo:
```diff
@@ -0,0 +1,126 @@
+  CDL.Continuous.Sources.Constant TOutBelowCutoff(final k=TOutCutoff - 2)
+    "Outdoor air temperature is slightly below the cutoff"
+    annotation (Placement(transformation(extent={{-40,140},{-20,160}})));
+  CDL.Continuous.Sources.Constant TOutCut(final k=TOutCutoff)
```
add comment

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/Validation/EconEnableDisableSingleZone_FreProSta_ZonSta.mo:
```diff
@@ -0,0 +1,126 @@
+  CDL.Continuous.Sources.Constant outDamPosMin(final k=0.1) "Minimal allowed economizer damper position"
+    annotation (Placement(transformation(extent={{-60,-160},{-40,-140}})));
+  CDL.Logical.Sources.Constant SupFanSta(k=true)
+    annotation (Placement(transformation(extent={{-160,-40},{-140,-20}})));
```
`SupFanSta` to be `supFanSta`, add comment

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Atomic/Validation/EconModulationMultiZone_TSup.mo:
```diff
@@ -3,29 +3,29 @@ model EconModulationMultiZone_TSup
-
-  CDL.Continuous.Sources.Constant outDamPosMin(k=0)
+  CDL.Continuous.Sources.Constant outDamPosMin(final k=0)
     annotation (Placement(transformation(extent={{-80,-20},{-60,0}})));
```
add comments

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/EconomizerMultiZone.mo:
```diff
@@ -7,40 +7,40 @@ model EconomizerMultiZone "Multiple zone VAV AHU economizer control sequence"
 
-  CDL.Interfaces.RealInput THeaSet(unit="K", quantity = "ThermodynamicTemperature")
+  CDL.Interfaces.RealInput THeaSet(final unit="K", final quantity = "ThermodynamicTemperature")
     "Supply air temperature heating setpoint" annotation (Placement(transformation(
```
move annotation to below line

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/EconomizerSingleZone.mo:
```diff
@@ -0,0 +1,202 @@
+  parameter Modelica.SIunits.VolumeFlowRate minVOut_flow=1.0 "Calculated minimum outdoor airflow rate";
+  parameter Modelica.SIunits.VolumeFlowRate desVOut_flow=2.0 "Calculated design outdoor airflow rate";
+  parameter Real minVOutMinFansSpePos(
+    final min=minVOutMaxFanSpePos, final max=desVOutMinFanSpePos, final unit="1") = 0.4
```
separate lines

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/EconomizerSingleZone.mo:
```diff
@@ -0,0 +1,202 @@
+    annotation (Placement(transformation(extent={{-140,70},{-120,90}}),
+        iconTransformation(extent={{-120,50},{-100,70}})));
+  CDL.Interfaces.RealInput uVOutMinSet_flow(final min=minVOut_flow, final max=desVOut_flow,
+    final unit="J/kg", final quantity="SpecificEnergy")
```
separate lines

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/Validation/EconomizerMultiZone_Disable.mo:
```diff
@@ -3,54 +3,56 @@ model EconomizerMultiZone_Disable
     annotation (Placement(transformation(extent={{-120,100},{-100,120}})));
-  CDL.Continuous.Sources.Constant TOutCut1(k=TOutCutoff)
+  CDL.Continuous.Sources.Constant TOutCut1(final k=TOutCutoff)
     annotation (Placement(transformation(extent={{-120,60},{-100,80}})));
```
add comment

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/Validation/EconomizerMultiZone_Disable.mo:
```diff
@@ -3,54 +3,56 @@ model EconomizerMultiZone_Disable
-  EconomizerMultiZone economizer(use_enthalpy=true) "Multizone VAV AHU economizer "
-    annotation (Placement(transformation(extent={{20,0},{40,20}})));
-  CDL.Logical.Sources.Constant fanStatus(k=true) "Fan is on"
+  CDL.Logical.Sources.Constant fanStatus(final k=true) "Fan is on"
```
`fanStatus` to be `fanSta`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/Validation/EconomizerMultiZone_Disable.mo:
```diff
@@ -3,54 +3,56 @@ model EconomizerMultiZone_Disable
+  CDL.Integers.Sources.Constant freProSta(final k=Constants.FreezeProtectionStages.stage0) "Freeze protection status is 0"
     annotation (Placement(transformation(extent={{-80,-130},{-60,-110}})));
-  CDL.Integers.Sources.Constant ZoneState(k=Constants.ZoneStates.heating) "Zone State is heating"
+  CDL.Integers.Sources.Constant zoneState(final k=Constants.ZoneStates.heating) "Zone State is heating"
```
`zoneState` to be `zonSta`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/Validation/EconomizerMultiZone_Disable.mo:
```diff
@@ -3,54 +3,56 @@ model EconomizerMultiZone_Disable
+  CDL.Integers.Sources.Constant zoneState(final k=Constants.ZoneStates.heating) "Zone State is heating"
     annotation (Placement(transformation(extent={{-80,-70},{-60,-50}})));
-  CDL.Integers.Sources.Constant OperationMode(k=Constants.OperationModes.occModInd) "AHU operation mode is Occupied"
+  CDL.Integers.Sources.Constant operationMode(final k=Constants.OperationModes.occModInd) "AHU operation mode is Occupied"
```
`operationMode` to be `opeMod`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/ASHRAE/G36/Composite/Validation/EconomizerSingleZone_Disable.mo:
```diff
@@ -0,0 +1,174 @@
+  CDL.Integers.Sources.Constant freProSta(final k=Constants.FreezeProtectionStages.stage0)
+    "Freeze protection stage is 0"
+    annotation (Placement(transformation(extent={{-80,-130},{-60,-110}})));
+  CDL.Integers.Sources.Constant ZoneState(final k=Constants.ZoneStates.heating) "Zone State is heating"
```
`ZoneState` tp be `zonSta`

milicag commented:
I can't seem to figure out the cause of[ this error](https://travis-ci.org/lbl-srg/modelica-buildings/jobs/259238663), since the model reported runs fine on my local machine. How do I access ```simulation.log``` of Travis runs?

mwetter commented:
@milicag 
If you set in Dymola
```
Advanced.PedanticModelica = true;
```
and check the model that fails, e.g.,
```
*** Error: Model check failed for 'Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.Atomic.Validation.EconEnableDisableSingleZone_FreProSta_ZonSta'.
```
then Dymola will report 
```
In class Buildings.Experimental.OpenBuildingControl.ASHRAE.G36.Atomic.Validation.EconEnableDisableSingleZone_FreProSta_ZonSta annotation, the annotation 'Placement, Placement' is unknown.
```
I corrected it in the commit above.

milicag commented:
@mwetter Thanks for the tip.

mwetter commented:
@milicag 
I reviewed all models and did some revision, mainly in the documentation. The main changes were that I removed the default values for `minVOut_flow` and `desVOut_flow` as we should only set default values where they are generally applicable, e.g., primarily non-dimensional quantities.

In some models, you used a timer, and a boolean "and" to test whether the input of the timer is true and the output of the timer is bigger than zero. I removed this test as the output of the timer is always bigger than zero if the input is true, except at the instant when the timer is activated.

Can you please address the following and then merge if all tests pass:
- [ ] In `Atomic.EconEnableDisable`, different default values are used for `smaDisDel` for the single and multi-zone. Please check if this is intentional, and if so, state it somewhere in the documentation or the code so we know why.
- [ ] In `Atomic.EconModulationSingleZone` and `Atomic.EconModulationMultiZone`, the figure in the info section shows `TCooSet` but the input connector of the block is `THeaSet`. I thought we discussed this already but can't find a separate issue.)
- [ ] In `EconomizerMultiZone` and `EconomizerSingleZone`, all parameters must be propagated. Otherwise, they will not be editable by the user through a gui menu. If the parameter list gets too long, you can use `annotation(Dialog(group="xxx"))` to group them under a common heading. Alternatively, parameters that are typically not to be changed could be done in a separate `tab`. We use tabs for those parameters that typically are not of interest to users except advanced users if we can provide a good default value. (Users should not have to open tabs to set all required parameters.)
- [ ] Please run the regression tests. After my changes, the results changed. I think they are due to the changes in control gain.

milicag commented:
@mwetter  This comment addresses the third task from your review. I've introduced the parameter groups (```Controller parameters```, ```Physical damper position limits```, ```Delay parameters at disable```). In the full economizer sequences I am omitting only those parameters propagated in the atomic sequences, which have values explicitly stated in G36. Here is an example from the damper position limit sequence:

![damperlimitsequence_parameterdefaults_distinctionbetweeng36defaultsandusereditableparams](https://user-images.githubusercontent.com/5322943/28985679-861ecd36-7918-11e7-8232-032fecacf95a.PNG)

I am assuming that the user who'd have a need to edit the controller signal limits will be able to drill down to the atomic sequence. The alternative is to give it a ```tab``` called something like ```G36 Defaults```. Let me know if you'd prefer that.

I'll push up my changes in a bit.

mwetter commented:
@milicag :  
When you group parameters in the dialog window, I would suggest not to repeat the word "parameter" as all of these settings are always parameters.
If you think a user has a need to edit parameters of a block that is contained in a sequence, then these need to be exposed so that users don't have to discover the lower level parameters and learn the syntax to assign them. Please use a different term than `G36 Defaults` because once the value is changed, these are not the default values anymore.

milicag commented:
@mwetter I'm using Parameter since there is no common Parameter title anywhere in the GUI, so that is a fact known only to Modelica users. I can remove it. I'll add a tab "Advanced Paramters", or simply "Advanced", whichever you prefer.

mwetter commented:
I would just call it "Advanced"

On Fri, Aug 4, 2017 at 4:31 PM, Milica Grahovac <notifications@github.com>
wrote:

> @mwetter <https://github.com/mwetter> I'm using Parameter since there is
> no common Parameter title anywhere in the GUI, so that is a fact known only
> to Modelica users. I can remove it. I'll add a tab "Advanced Paramters", or
> simply "Advanced", whichever you prefer.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lbl-srg/modelica-buildings/pull/855#issuecomment-320374305>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRHuHmDq6SXnx0Mj49jjO80L2KyevIGks5sU6nMgaJpZM4Of6tR>
> .
>

milicag commented:
@mwetter The new ```TiMod``` = 30 will not demonstrate the return air damper limit modulation in the test. One option is to have a longer period of smaller delta between the setpoint and the measurement in the test.

mwetter commented:
@milicag Then it is better to run the test over a longer time horizon. We often see users copying models from examples, and having a very small time constant will lead to unstable control if used in a more realistic setting.

milicag commented:
- [] add ```Advanced``` tab for advanced parameters (propagate all remaining parameters from atomic sequences)

milicag commented:
@mwetter I've addressed all of your comments. In addition, I protected blocks in some of the validation models and introduced several formatting improvements. Most significant update is the introduction of parameter tabs. Please let me know if you'd like any further changes.
